src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -116,6 +116,23 @@ extern "C" void QCALLTYPE RuntimeTypeHandle_CreateInstanceForAnotherGenericParam
     END_QCALL;
 }
 
+extern "C" void QCALLTYPE RuntimeTypeHandle_InternalAlloc(MethodTable* pMT, BOOL check, QCall::ObjectHandleOnStack allocated)
+{
+    QCALL_CONTRACT;
+
+    _ASSERTE(pMT != NULL);
+
+    BEGIN_QCALL;
+
+    GCX_COOP();
```

Comment:
I don't believe so. The `SuppressGCTransition` is designed to avoid costs and suppress the GC transition. I'm not sure what would need to change, but at a minimum the `BEGIN_QCALL` would likely need to updated and we'd need to understand the different contract. Regardless of the outcome changing this one instance when the pattern is used in many places isn't the point of this PR.

I defer to @jkotas or @janvorli if this is advisable or even possible.

Reply:
> Could these QCalls use [SuppressGCTransition] since they immediately switch to cooperative mode?

SuppressGCTransition can be only used in situations described in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.suppressgctransitionattribute?view=net-9.0#remarks . This call is not one of those. For example, it can throw exceptions or it can do synchronization as part of the GC.

We need a full transition here.

In theory, we can introduce a new special `FullGCTransitionWithoutSwitchToPreemptiveMode` that would micro-optimize a few instructions in the QCalls that immediately switch to cooperative mode. We have not seen evidence that it is worth the complexity.

Reply:
Thanks, I didn't know that a GC transition does more things besides switching to preemptive mode.